src/libs/elektra/keyset.c

```diff
@@ -975,6 +1094,78 @@ ssize_t ksCopyInternal (KeySet * ks, size_t to, size_t from)
 	return ret;
 }
 
+/**
+ * Searches for the start and optionally end of the key hierachy rooted at @p root in @p ks.
+ *
+ * The main use-case for this function is this kind of loop:
+ *
+ * @code{.c}
+ * elektraCursor end;
+ * for (elektraCursor it = ksFindHierarachy (ks, root, &end); it < end; ++it)
+ * {
+ * 	Key * cur = ksAtCursor (ks, it);
+ * }
+ * @endcode
+ *
+ * @param ks   The keyset to search in
+ * @param root The root of the hierachy to find
+ * @param end  If this is not NULL, it will be set to position of the first
+ *             key after @p root that is not below @p root. This is useful
+ *             for loops like the one above.
+ *             If not keys below @p root exist in @p ks, @p end will always
+ *             be set to the size of @p ks. This way a loop like the one above
+ *             will still work correctly.
+ *
+ * @retval -1 if @p ks or @p root are NULL
+ * @return the position of either @p root itself or the first key below
+ *         @p root that is part of @p ks. If no keys below @p root exist
+ *         in @p ks, the size of @p ks is returned. The snippet above
+ *         shows why this is useful.
+ */
+elektraCursor ksFindHierarchy (const KeySet * ks, const Key * root, elektraCursor * end)
```

Comment:
I found this neat alternative to the `ksExtract` function we discussed in #2992. I put it into `kdbprivate.h` for now, but I really think it should be in `kdb.h` since it will be incredibly useful.

I'll also rewrite `elektraKsFindCutpoint` (used by `ksCut`) to use this, instead of just looking for the root and then looping over the following keys.